File: src/main/java/com/github/hirsivaja/ip/ipv6/Ipv6Packet.java
package com.github.hirsivaja.ip.ipv6;

import com.github.hirsivaja.ip.IpPacket;

import java.nio.ByteBuffer;

public record Ipv6Packet(Ipv6Header header, Ipv6Payload payload) implements IpPacket {

    @Override
    public void encode(ByteBuffer out) {
        header.encode(out);
        payload.encode(out);
    }

    @Override
    public int length() {
        return header.length() + payload.length();
    }

    public static IpPacket decode(ByteBuffer in) {
        return decode(in, true);
    }

    public static IpPacket decode(ByteBuffer in, boolean ensureChecksum) {
        Ipv6Header header = Ipv6Header.decode(in, ensureChecksum);
        byte[] payloadBytes = new byte[header.payloadOnlyLength()];
        in.get(payloadBytes);
        ByteBuffer payloadBuffer = ByteBuffer.wrap(payloadBytes);
        Ipv6Payload payload = Ipv6Payload.decode(payloadBuffer, ensureChecksum, header);
        return new Ipv6Packet(header, payload);
    }
}

File: src/main/kotlin/com/github/hirsivaja/ip/packet/Ipv6Packet.kt
package com.github.hirsivaja.ip.packet

import com.github.hirsivaja.ip.packet.header.Ipv6Header
import com.github.hirsivaja.ip.packet.payload.Ipv6Payload
import java.nio.ByteBuffer

data class Ipv6Packet(override val header : Ipv6Header, override val payload : Ipv6Payload) : IpPacket {
    companion object {
        fun decode(buf: ByteBuffer, ensureChecksum: Boolean): IpPacket {
            val header = Ipv6Header.decode(buf)
            val payload = Ipv6Payload.decode(buf, ensureChecksum, header)
            return Ipv6Packet(header, payload)
        }
    }
}
